### PR TITLE
Make default_library per-subproject built-in option

### LINKED ISSUE #4155: Can't use `-Dsubproject:option` format for `default_library`
Running `meson -Dgstreamer:default_library=static` yields:
```
|WARNING: In subproject gstreamer: Unknown options: "gstreamer:default_library"
```

I thought this already worked. @xclaesse, any ideas? :)

### PR DESCRIPTION
**Pull request by xclaesse**
Closes #4155.

### CODE REVIEW

**Comment by jpakkane** [comment]
The obvious followup to this is "well why is not [every other option X] _also_ not a per-project option"? And since that is a fairly good question, it leads us directly to #3001.

**Comment by textshell** [comment]
I think the use case for #3001 is different. 3001 is for the user / packager / integrator to allow doing overrides.

I think this is more for use cases where you want something from wrapdb but for one reason or another this needs to be statically build (maybe with hidden visibility) into the toplevel project where the toplevel project still allows it self to be build as shared library. Use cases like "we use that json library internally but can neither depend on this to be available nor are allowed to leak these symbols out into our users."

**Comment by jpakkane** [comment]
But that's exactly the use case that #3001 provides for. In JSON pseudo-notation you would have an override file with this content:

```json
{
  "target_type_overrides": [
    {
      "subproject": "json",
      "target_name": "jsonlibrarynamehere",
      "target_type": "static_library"
    }
  ]
}
```

And it would cause the json library target to be built as a static library regardless of how it is specified in the build files.

**Comment by tp-m** [comment]
Not sure I understand. As a project maintainer, if we fall back to a subproject/wrap I may want to specify in the build files using `default_options:` that the subproject should be built statically regardless of the global project setting. I was highly surprised to find that this not only doesn't work, but there's no warning or anything about that either. How would the json stuff fit into this? Why would I have to provide an override file instead of putting stuff where it belongs in this case namely in the build file? I might have misunderstood your proposal.

**Comment by nirbheek** [comment]
A concrete example of this is that GStreamer builds ffmpeg as a subproject and that needs to be built statically in the default configuration because the library is linked into the GStreamer plugin for ffmpeg. However, ffmpeg itself would not default to a static build.

**Comment by xclaesse** [comment]
It makes sense for the master project to decide to link static or dyn on its subproject, rather than leaving the decision to user.

**Comment by nirbheek** [comment]
It also makes sense for the user to decide that if they want a static build of a plugin dependency, say, libpng, and that uses the same framework as setting the `default_options:` inside `dependency()` in the build file. A generic overriding behaviour would be useful, but it's a larger feature set.

**Comment by jpakkane** [comment]
Language standard (especially for C) is the obvious one, it has been requested multiple times. You might also want to have optimization enabled for dependency subprojects but not the master project for easy debuggability. And so on.

> this PR adds all the infrastructure for having more per-subproject builtin options

That is exactly what I don't want to happen. Adding them one by one leads to the case that half of options can be set per subproject and half can't without any rhyme or reason. It is better to have one good solution for this.

**Comment by xclaesse** [comment]
>  It is better to have one good solution for this.

Why isn't this PR the one good solution?

**Comment by xclaesse** [comment]
Reasons I think this PR should be accepted:
1) It exists and can be merged today. #3001 has been there for months and nobody wrote the first line of a patch.
2) For the `c_std` and `default_library` the decision often comes from the project itself, and not the user. We should not require user to write some complicated json (?) file when it could just work out of the box.
3) It is simple and targetted for real use cases, not something vague that tries to fix all problems at once.

**Comment by xclaesse** [comment]
> Adding them one by one leads to the case that half of options can be set per subproject and half can't without any rhyme or reason. 

Actually, you are right. Instead of thinking which option should be per-subproject, we could make them all per-subproject except for a few like dirs.

**Comment by TheQwertiest** [comment]
> Instead of thinking which option should be per-subproject, we could make them all per-subproject except for a few like dirs.

Any chance for this feature/PR to be (re)implemented?

File: mesonbuild/coredata.py [hunk 1]
```diff
@@ -662,11 +678,13 @@ def set_default_options(self, default_options, subproject, env):
...
         # Set default options as if they were passed to the command line.
-        # Subprojects can only define default for user options.
+        # Subprojects can only define default for user options and not yielding
+        # builtin option.
         from . import optinterpreter
         for k, v in default_options.items():
             if subproject:
-                if optinterpreter.is_invalid_name(k, log=False):
+                if (k not in builtin_options or builtin_options[k].yielding) \
+                      and optinterpreter.is_invalid_name(k, log=False):
```
Comment (ignatenkobrain) [inline mesonbuild/coredata.py]: __[Flake8]__ [E126] continuation line over-indented for hanging indent ([link](https://sider.review/gh/repos/19784232/pulls/4157#issue-4358606))

<sub>posted by Sider</sub>

**Comment by xclaesse** [comment]
@TheQwertiest updated this PR, but it's not working yet for compiler arguments. It's hard to change every places we try to get c_args and friends.

File: mesonbuild/mconf.py [hunk 2]
```diff
@@ -43,6 +43,11 @@ def __init__(self, build_dir):
             self.build_dir = os.path.dirname(self.build_dir)
         self.build = None
         self.max_choices_line_length = 60
+        self.name_col = []
+        self.value_col = []
+        self.choices_col = []
+        self.descr_col =  []
```
Comment (ignatenkobrain) [inline mesonbuild/mconf.py]: __[Flake8]__ [E222] multiple spaces after operator ([link](https://sider.review/gh/repos/19784232/pulls/4157#issue-4377388))

<sub>posted by Sider</sub>

**Comment by jpakkane** [comment]
Thinking about this further, we probably need to support at least these two use cases:

- a `c_std` set on the master project propagates to all subprojects by default
- a subproject should have a way of explicitly saying "no, I want gnu11 instead of whatever" or "warning level is only 1 for this project"

If you have this, do you need to set project options per subproject from the outside?

**Comment by TheQwertiest** [comment]
@jpakkane , `c_std` is not the only option that needs such handling though: `default_library`, `optimization`, `werror`, `warning_level` and probably some of the `b_*` options need this as well.

**Comment by xclaesse** [comment]
For c_std I don't really have a use case for defining it from outside. But for default_library I think it makes sense to let the user decide. For example with ffmpeg it has licensing implications.

**Comment by xclaesse** [comment]
The obvious next candidates to be per-subproject are `werror`and `warning_level`, and I think for them it makes sense to define them from the outside. For example if a newer gcc version start producing more warnings I could want to disable `werror` for some subprojects dependencies that are outside of my control but keep it for the master project on which I'm working and actually want to catch&fix those warnings.

**Comment by xclaesse** [comment]
Extracted the configure output refactoring to its own PR: https://github.com/mesonbuild/meson/pull/6014

File: mesonbuild/coredata.py [hunk 3]
```diff
@@ -696,11 +713,13 @@ def set_default_options(self, default_options, subproject, env):
...
```
Comment (ignatenkobrain) [inline mesonbuild/coredata.py]: __[Flake8]__ continuation line over-indented for hanging indent ([see on Sider](https://sider.review/gh/repos/19784232/pulls/4157#issue-4884709))

<details>

| Rule |
| - |
| E126 |

</details>
<sub>posted by Sider</sub>

**Comment by xclaesse** [comment]
kept here only the patch to make `default_library` per-subproject. Others are a bit more complicated and can go in their own PR later.

**Comment by TheQwertiest** [comment]
Good call (probably). I really hope though, that warning level and cpp standard will be next ones to be tackled.

File: mesonbuild/mintro.py [hunk 4]
```diff
@@ -178,9 +178,10 @@ def list_targets(builddata: build.Build, installdata, backend: backends.Backend)
     return tlist
 
 def list_buildoptions_from_source(intr: IntrospectionInterpreter) -> List[Dict[str, Union[str, bool, int, List[str]]]]:
-    return list_buildoptions(intr.coredata)
+    subprojects = [i['name'] for i in intr.project_data['subprojects']]
+    return list_buildoptions(intr.coredata, subprojects)
 
-def list_buildoptions(coredata: cdata.CoreData) -> List[Dict[str, Union[str, bool, int, List[str]]]]:
+def list_buildoptions(coredata: cdata.CoreData, subprojects : List[str] = None) -> List[Dict[str, Union[str, bool, int, List[str]]]]:
```
Comment (ignatenkobrain) [inline mesonbuild/mintro.py]: __[Flake8]__ whitespace before ':' ([see on Sider](https://sider.review/gh/repos/19784232/pulls/4157#issue-4973892))

<details>

| Rule |
| - |
| E203 |

</details>
<sub>posted by Sider</sub>

**Comment by emersion** [comment]
Any news about this? What needs to be done here?

**Comment by xclaesse** [comment]
Blocking in someone to do the review.

**Comment by TheQwertiest** [comment]
@jpakkane , any possibility for this PR to get reviewed/merged?

File: mesonbuild/coredata.py [hunk 5]
```diff
@@ -367,7 +367,8 @@ def __init__(self, options: argparse.Namespace, scratch_dir: str):
         self.install_guid = str(uuid.uuid4()).upper()
         self.target_guids = {}
         self.version = version
-        self.init_builtins()
+        self.builtins = {} # : Dict[str, UserOption]
```
Comment (dcbaker) [inline mesonbuild/coredata.py]: The type here should be `OptionDictType`

File: mesonbuild/coredata.py [hunk 6]
```diff
@@ -491,15 +493,25 @@ def sanitize_dir_option_value(self, prefix, option, value):
             value = value[skip:]
         return value
 
-    def init_builtins(self):
+    def init_builtins(self, subproject):
```
Comment (dcbaker) [inline mesonbuild/coredata.py]: subproject is a `str`, right?

File: mesonbuild/coredata.py [hunk 7]
```diff
@@ -943,17 +962,19 @@ class BuiltinOption(Generic[_T, _U]):
     Currently doesn't support UserIntegerOption, or a few other cases.
     """
 
-    def __init__(self, opt_type: Type[_U], description: str, default: Any, yielding: Optional[bool] = None, *,
+    def __init__(self, opt_type: Type[_U], description: str, default: Any, yielding: Optional[bool] = True, *,
```
Comment (dcbaker) [inline mesonbuild/coredata.py]: Is yielding really optional anymore?

File: mesonbuild/coredata.py [hunk 8]
```diff
@@ -943,17 +962,19 @@ class BuiltinOption(Generic[_T, _U]):
...
                  choices: Any = None):
         self.opt_type = opt_type
         self.description = description
         self.default = default
         self.choices = choices
         self.yielding = yielding
 
-    def init_option(self) -> _U:
+    def init_option(self, value=None) -> _U:
```
Comment (dcbaker) [inline mesonbuild/coredata.py]: It would be nice to have a type annotation for value

File: mesonbuild/mintro.py [hunk 9]
```diff
@@ -178,9 +178,10 @@ def list_targets(builddata: build.Build, installdata, backend: backends.Backend)
     return tlist
 
 def list_buildoptions_from_source(intr: IntrospectionInterpreter) -> List[Dict[str, Union[str, bool, int, List[str]]]]:
-    return list_buildoptions(intr.coredata)
+    subprojects = [i['name'] for i in intr.project_data['subprojects']]
+    return list_buildoptions(intr.coredata, subprojects)
 
-def list_buildoptions(coredata: cdata.CoreData) -> List[Dict[str, Union[str, bool, int, List[str]]]]:
+def list_buildoptions(coredata: cdata.CoreData, subprojects: List[str] = None) -> List[Dict[str, Union[str, bool, int, List[str]]]]:
```
Comment (dcbaker) [inline mesonbuild/mintro.py]: The type is wrong, you can't assign a None to a List[str], I think this should be Optional[List[str]]

File: docs/markdown/Reference-manual.md [hunk 10]
```diff
@@ -1546,7 +1546,8 @@ arguments:
    that override those set in the subproject's `meson_options.txt`
    (like `default_options` in `project`, they only have effect when
    Meson is run for the first time, and command line arguments override
-   any default options in build files)
+   any default options in build files). *Since 0.53.0* `default_library`
```
Comment (dcbaker) [inline docs/markdown/Reference-manual.md]: Should be 0.54.0 now

File: docs/markdown/snippets/per_subproject_builtin.md [hunk 11]
```diff
@@ -0,0 +1,15 @@
...
+
+Most of the time this would be used either by the parent project by setting
+subproject's default_options (e.g. `subproject('foo', default_options: 'default_library=static')`),
+or by the user using the command line `-Dfoo:default_library=static`.
+
+The value is overriden in this order:
+- Value from parent project
+- Value from subproject's default_options if set
+- Value from subproject() default_options if set
+- Value from command line if set
```
Comment (dcbaker) [inline docs/markdown/snippets/per_subproject_builtin.md]: This means that command line has the highest priority, right?

Comment (xclaesse) [reply docs/markdown/snippets/per_subproject_builtin.md]: Yes, that list reads as each item gets overridden by the next. So command line is the ultimate.

**Comment by jpakkane** [comment]
Regardless of everything else, the solution chosen must cater for both `c_std` as well as default library.

File: mesonbuild/coredata.py (hunk 7 above)
Comment (xclaesse) [reply mesonbuild/coredata.py]: It's always bool now indeed. Fixed.

File: mesonbuild/coredata.py (hunk 8 above)
Comment (xclaesse) [reply mesonbuild/coredata.py]: Can be any type, depending on which UserOption subclass it is. value is not annotated in other places.

File: docs/markdown/Reference-manual.md (hunk 10 above)
Comment (xclaesse) [reply docs/markdown/Reference-manual.md]: Done.

File: mesonbuild/mintro.py (hunk 9 above)
Comment (xclaesse) [reply mesonbuild/mintro.py]: Done.

File: mesonbuild/coredata.py (hunk 6 above)
Comment (xclaesse) [reply mesonbuild/coredata.py]: Yes. Done.

File: mesonbuild/coredata.py (hunk 5 above)
Comment (xclaesse) [reply mesonbuild/coredata.py]: Need to fix other dict too then. Done.

File: mesonbuild/coredata.py (hunk 8 above)
Comment (xclaesse) [reply mesonbuild/coredata.py]: @dcbaker I'm not familiar with annotation system, what should I put here?

**Comment by xclaesse** [comment]
> Regardless of everything else, the solution chosen must cater for both c_std as well as default library.

This PR used to do that as well, but it's more complicated and IIRC it had issues. So I decided to keep only this part for now, more should be added later.

@dcbaker rebased the branch and fixed your comments.

**Comment by jpakkane** [comment]
The merge rules say that these sort of changes require the project lead's approval, which I very specifically have not given. If the issues I have raised are not addressed, I reserve the right to revert this.

**Comment by xclaesse** [comment]
I fixed all review comments. You did not voice any objections in months. It is intentional to make changes incrementally, it already though insanely long to get a review, if I include everything at once it will never happen.

**Comment by eli-schwartz** [comment]
???

> Regardless of everything else, the solution chosen must cater for both c_std as well as default library.

This certainly looks like an objection to me.

**Comment by xclaesse** [comment]
I don't think it is fair to come with such objection 1 YEAR after this PR has been submitted and it got finally approved by another main contributor. Especially since I explained why I splitter this PR 2 months ago. At some point things need to move forward.

**Comment by eli-schwartz** [comment]
So your position has moved, and instead of claiming there is no objection, you're claiming the objection isn't a fair objection?

I'm merely pointing out here that I certainly understand why @jpakkane might feel this was merged in bad faith, and needs to be reverted by project leader dictatorial authority.

It seems like a sad situation for a project developer to pick a fight with the project leader over something that can only end badly, and poisons the well against the feature existing at all.

**Comment by xclaesse** [comment]
There are no justified objections. If some arguments other than authority were voice I wouldn't have merged this.

**Comment by xclaesse** [comment]
For reference, I pushed the WIP code I've got so far for making compiler options per subproject too: https://github.com/mesonbuild/meson/pull/6729.

**Pull request merged**